### ISSUE TITLE
Strainer > Fuzztester

### ISSUE THREAD

**Issue by cookiengineer** [Artificial-Engineering/lycheejs#181]
The new `/libraries/strainer` needs a new fuzz tester that is based on the new API definition format located in the project's or library's `/api/Definition.json` file.

This file format is given and ready to be used for the ahead-of-time compiler in future. In order to have more typedef tests, it is important to implement all data types and to fuzz the APIs in order to make sure that the Definitions won't break given different (unexpected) data types as parameters or settings.

Necessary parts to implement:

- [ ] Test Suite API
- [ ] Test Definition API (`new strainer.fuzz.Simulation({ id: "lychee.Input" })`)
- [ ] `./bin/strainer-fuzzer.js` that runs as both daemon and sub-process
- [ ] low-level data types (`Null`, `Undefined`, `Boolean`, `Number`, `RegExp`, `String`, `Array`, `Object`)
- [ ] lychee data types (`Buffer`, `Config`, `Font`, `Music`, `Sound`, `Texture`, `Stuff`)
- [ ] high-level data types (`settings` object)
- [ ] implement testing API
- [ ] generate testing code and write it to `/libraries/<library>/path/Definition.js` so it can be run manually again

**Comment by cookiengineer** [comment]
As of https://github.com/Artificial-Engineering/lycheejs/commit/a8d988ba40064f7acafeaf88753e5ecb98b0eec4 the strainer has now the necessary featureset to trace all static and dynamic memory parts.

All API plugins have been updated and refactored to reflect the newer, better JSON structure that allows to trace private and public memory parts, mutations of variables and even mutations of return values etc.

The next todo here is to add a fuzztesting library (suite / plugin) that will leverage the JSON structure to mutate the parameters and values into evil data structures.

**Comment by cookiengineer** [comment]
The static code analysis got much better and now evaluates all function return values and traces their mutations. So as of today, void APIs are forbidden across the whole stack, except for `update()` and `render()` due to performance reasons.

The `strainer/source/api/Composite.js` and `strainer/source/api/Module.js` need to further throw errors when parameters were untraceable. Currently, `val.type === 'undefined'` is allowed for parameters, which is not so good for the upcoming fuzztester.

Personal note: [./libraries/strainer/source/api/Composite.js#L366] is the point where parameters have to be validated against their types and mutations.

The idea of the `strainer check` command will now be static code intelligence, whilst `strainer stage` will represent dynamic code intelligence (and therefore fuzztesting).

**Comment by cookiengineer** [comment]
Personal note: The `strainer/source/api/PARSER.js` currently seems to not correctly parse `/libraries/lychee/verlet/Entity.js#setPosition`.

The `instanceof _Vector3` usage is correct and should've been mapped correctly, but it isn't. This seems to be probably a bug of the `mutations()` or `parameters()` tracer which should also use the `val.chunks` notation so that the `api/Composite.js` is able to fill it with the static memory knowledge (as `_Vector3` is already in memory at that point).

**Comment by cookiengineer** [comment]
The `lycheejs-strainer` now parses correctly all data types, and allows a lookup for the memory when it was prefixed with a `_` (underscore).

Also, events, methods, parameters, values are all now dynamically traced via both memory lookups and mutations. This now generates a fully prepared JSON API file for the fuzz tester, so that it can build proper event and flow graph simulations.

**Comment by cookiengineer** [comment]
After the code review, the only thing that's really bothering me implementation-wise is the handling of the `deserialize()` and `serialize()` method.

On one hand the signatures would be nice to be parsed out failsafe (e.g. with type == 'SerializationBlob'), but on the other hand this doesn't make any sense storage-wise; as every serialize/deserialize method has identical signatures.

So I guess this is okay, as it allows dynamic memory lookup in case the serialization interface was inherited (e.g. value == '_Emitter.prototype.serialize.call(this')) - so that other tools in future can lookup the composite chain dynamically and at runtime.

**Comment by cookiengineer** [comment]
As of today, the dynamic memory lookup has been implemented in the `strainer check` flow. This means that the fuzzer is now unnecessary - as everything is trace-able at compile time and doesn't require a live testing time.

The fuzzer will only be necessary in the case of learning; so it will probably be now a part of the studio project and not the strainer project. This issue will be reopened once #187 is completed and implemented in a proper way.

There might be a way that even in that case the fuzzer will never be necessary; as the strainer will now prevent dynamically typed methods, properties, enums and events anyways - and its knowledge index contains all relevant data that is necessary for fuzzy suggestions.

In short: Fuzzer is unnecessary because of dynamic memory tracing algorithm.
